File: backend/app/models/modal.py
from __future__ import annotations

import uuid

from datetime import time
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    Time,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import BaseModel


class EmployeeModal(BaseModel):
    """Employee modal analysis — transport mode preferences and mobility profile."""

    __tablename__ = "employee_modal"
    __table_args__ = (
        UniqueConstraint("employee_id", name="uq_employee_modal_employee_id"),
        Index("idx_modal_employee", "employee_id"),
    )

    employee_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("employee.id", ondelete="CASCADE"),
        nullable=False,
    )

    # Transport modes
    primary_mode: Mapped[str] = mapped_column(String(50), nullable=False)
    alternative_mode: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Commute metrics
    distance_km: Mapped[Decimal | None] = mapped_column(Numeric(8, 2), nullable=True)
    travel_time_min: Mapped[int | None] = mapped_column(Integer, nullable=True)
    frequency: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Interest & reasoning
    interest_company_transport: Mapped[str | None] = mapped_column(
        String(30), nullable=True
    )
    reason_current_mode: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Schedule
    departure_time: Mapped[time | None] = mapped_column(Time, nullable=True)

    # Pickup preferences
    accepts_common_pickup: Mapped[bool] = mapped_column(
        Boolean, server_default="true", nullable=False
    )
    max_pickup_distance_meters: Mapped[int] = mapped_column(
        Integer, server_default="500", nullable=False
    )

    # Vehicle & carpooling
    has_private_car: Mapped[bool] = mapped_column(
        Boolean, server_default="false", nullable=False
    )
    volunteer_driver: Mapped[bool] = mapped_column(
        Boolean, server_default="false", nullable=False
    )
    carpool_seats_available: Mapped[int] = mapped_column(
        Integer, server_default="0", nullable=False
    )
    max_detour_minutes: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # Opt-in & notes
    bonus_opt_in: Mapped[bool] = mapped_column(
        Boolean, server_default="false", nullable=False
    )
    observations: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Relationships
    employee: Mapped[Employee] = relationship("Employee", lazy="selectin")

File: frontend/src/types/modal.ts
export interface EmployeeModal {
  id: string;
  employee_id: string;
  employee_name: string | null;
  primary_mode: string;
  alternative_mode: string | null;
  distance_km: number | null;
  travel_time_min: number | null;
  frequency: string | null;
  interest_company_transport: string | null;
  accepts_common_pickup: boolean;
  max_pickup_distance_meters: number;
  has_private_car: boolean;
  volunteer_driver: boolean;
  carpool_seats_available: number;
  max_detour_minutes: number | null;
  bonus_opt_in: boolean;
  observations: string | null;
}

export interface EmployeeModalUpsert {
  primary_mode: string;
  alternative_mode?: string | null;
  distance_km?: number | null;
  travel_time_min?: number | null;
  frequency?: string | null;
  interest_company_transport?: string | null;
  accepts_common_pickup?: boolean;
  max_pickup_distance_meters?: number;
  has_private_car?: boolean;
  volunteer_driver?: boolean;
  carpool_seats_available?: number;
  max_detour_minutes?: number | null;
  bonus_opt_in?: boolean;
  observations?: string | null;
}

export interface ModalDistribution {
  mode: string;
  count: number;
  percentage: number;
}

export interface ModalSiteDistribution {
  site_id: string;
  site_name: string;
  distribution: ModalDistribution[];
}

export interface ModalStats {
  total: number;
  distribution: ModalDistribution[];
  by_site: ModalSiteDistribution[];
}

export interface ShiftModalData {
  shift_time: string;
  total: number;
  distribution: ModalDistribution[];
}

export interface MobilityScore {
  employee_id: string;
  employee_name: string;
  score: number;
  factors: Record<string, number>;
}
